doc/threading_notes.md

```diff
@@ -0,0 +1,49 @@
+# Threading model and optimizing for resource constrained devices
+
+## IoTHub Device Client and the \_LL\_ (lower layer) API
+
+### The lower layer
+
+There are two distinct but closely related sets of APIs for the IoTHub client.
+Applications that want to run on a single thread should use the \_LL\_ (aka lower level) API.  Sometimes the hardware/OS can only support one thread at a time or else spinning additional threads is expensive.  Applications on more capable hardware may also use the \_LL\_ layer for greater control over scheduling.
+
+Calls to the IoThub API's ending in Async do not cause the operation to execute immediately.  Instead, the work - both for sending data across the network and receiving data - is queued.  The application must manually schedule the work by calling `IoTHubDeviceClient_LL_DoWork` (when using a device client).
+
+This is typically done in some sort of simple loop, e.g.
+
+```c
+    IOTHUB_DEVICE_CLIENT_LL_HANDLE iotHubLLHandle;
+
+    while (appIsRunning) {
+        SendTelemetryIfThereIsAnyToSend(iotHubLLHandle);
+        IoTHubDeviceClient_LL_DoWork(iotHubLLHandle);
+        Sleep(1000);
+    }
+```
+
+**CRITICAL(!) notes for \_LL\_ application development**  
+*NOT FOLLOWING THESE STEPS WILL POTENTIALLY RESULT IN VERY HARD TO DIAGNOSE BUGS*
+
+* You **MUST NOT** use an \_LL\_ handle on more than one thread (unless you provide *all* your own locking).  Part of what makes the \_LL\_ layer use fewer resources is that it does not use locks.  It assumes a single threaded environment where locks are not needed.
+* Even if you are not actively sending data while using \_LL\_ layer, you still need to periodically invoke the appropriate `DoWork` to make the client accept data from the network.  This periodic `DoWork` call is also required as otherwise the underlying network transport and/or service may timeout the service due to inactivity.
+
+### The convenience layer
+
+Applications may also use the non \_LL\_ API set, which is also known as the "convenience layer."  
```

Comment:
period before closing quotes

Reply:
I always taught - and some quick research seems to back this up - that the period goes in the quote.  E.G. https://style.mla.org/the-placement-of-a-comma-or-period-after-a-quotation/ for US at least.